fix sql-server dbm error in certain environments

it-ito opened the pull request:
### What does this PR do?
<!-- A brief description of the change being made with this pull request. -->
This is a pull request to change `select plan_handle, statement_start_offset , statement_end_offset` in SQLServer's DBM.
The code added in https://github.com/DataDog/integrations-core/pull/12613 return different results in some environments and an error occur.

ForExample:
- query
```sql
DECLARE @statement_start_offset int  = 14730;
DECLARE @statement_end_offset int = 14838;

SELECT
@statement_start_offset AS original_statement_start_offset,
CONVERT(int, CONVERT(varbinary(4),
    SUBSTRING(
        CONCAT(
        CONVERT(binary(4), @statement_start_offset),
        CONVERT(binary(4), @statement_end_offset) 
        ), 1, 4)
)) AS converted_statement_start_offset,
@statement_end_offset AS original_statement_end_offset,
CONVERT(int, CONVERT(varbinary(4),
    SUBSTRING(
        CONCAT(
        CONVERT(binary(4), @statement_start_offset),
        CONVERT(binary(4), @statement_end_offset)
        ),5, 4)
)) AS converted_statement_end_offset;
SELECT collation_name FROM sys.databases where database_id = DB_ID();
```
- returns
    - Collation : SQL_Latin1_General_CP1_CI_AS
    - ![collation_SQL_Latin1_General_CP1_CI_AS](https://user-images.githubusercontent.com/46398493/235837565-1a7555f4-cfc9-4128-b82c-44e2145f50f6.PNG)
    - Collation : Japanese_CI_AS
![collation_Japanese_CI_AS](https://user-images.githubusercontent.com/46398493/235837574-1c41fbc0-2ff7-443e-8acb-7239cfa56921.PNG)

### Motivation
<!-- What inspired you to submit this pull request? -->
this issue
- https://github.com/DataDog/integrations-core/issues/14006

I can't update datadog agent until fix this error

### Additional Notes
<!-- Anything else we should know when reviewing? -->

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached
- [ ] If the PR doesn't need to be tested during QA, please add a `qa/skip-qa` label.

it-ito commented on sqlserver/datadog_checks/sqlserver/statements.py:
```diff
@@ -97,20 +94,13 @@
-            CONVERT(binary(64), plan_handle),
-            CONVERT(binary(4), statement_start_offset),
-            CONVERT(binary(4), statement_end_offset))) as plan_handle_and_offsets,
+        max(plan_handle) as plan_handle,
+        max(statement_start_offset) as statement_start_offset,
+        max(statement_end_offset) as statement_end_offset,
```
Thanks, you are right. I forgot `query_hash` and `query_plan_hash` are the same, but not necessarily plan_handle.
This fix is wrong.

nenadnoveljic commented:
It seems like the change broke some tests:

```
=================================== FAILURES ===================================
________ test_statement_metadata[metadata1-expected_metadata_payload1] _________
tests/test_statements.py:482: in test_statement_metadata
    assert sample['db']['metadata']['comments'] == expected_metadata_payload['comments']
E   AssertionError: assert ['-- Test comment'] == None
------------------------------ Captured log call -------------------------------
WARNING  datadog_checks.base.checks.base.sqlserver:sqlserver.py:189 Autodiscovery is disabled, autodiscovery_include and autodiscovery_exclude will be ignored
WARNING  datadog_checks.base.checks.base.sqlserver:core.py:247 QueryManager initialized with no query
______________________ test_statement_basic_metrics_query ______________________
tests/test_statements.py:678: in test_statement_basic_metrics_query
    assert matching, "the test query should be visible in the query stats"
E   AssertionError: the test query should be visible in the query stats
E   assert []
- generated xml file: /home/runner/work/integrations-core/integrations-core/sqlserver/.junit/test-unit-py3.8-linux-2017-single$TOX_ENV_NAME.xml -
=========================== short test summary info ============================
FAILED tests/test_statements.py::test_statement_metadata[metadata1-expected_metadata_payload1] - AssertionError: assert ['-- Test comment'] == None
FAILED tests/test_statements.py::test_statement_basic_metrics_query - AssertionError: the test query should be visible in the query stats
assert []
============ 2 failed, 192 passed, 20 skipped in 322.33s (0:05:22) =============
```

it-ito commented:
@nenadnoveljic 
When there is a cache of queries with the same `query_hash` and `query_plan_hash`, [this max(plan_handle_and_offsets)](https://github.com/DataDog/integrations-core/blob/adf5d2a523ca41945747eb858c092a0d39523b01/sqlserver/datadog_checks/sqlserver/statements.py#L68)  may be different from before.
What do you think of the fix commit (ead85416daec947c7262cf16b78c8f6f3aa12e82) that would keep this from changing?

If the possibility of changing the results is acceptable, the test case can be modified.
`tests/test_statements.py::test_statement_basic_metrics_query` is affected by other tests, so if `DBCC FREEPROCCACHE` is available, this case would be a success.
But I don't know how to fix `tests/test_statements.py::test_statement_metadata[metadata1-expected_metadata_payload1]` .

nenadnoveljic commented:
Thank you for the changes!

I compared the performance to the old query. The new query performs better on my test database - lots of logical reads from the Worktable are eliminated:

Old query run 1:
```
Started executing query at Line 5
SQL Server parse and compile time: 
   CPU time = 44 ms, elapsed time = 44 ms.
(26 rows affected)
Table 'Worktable'. Scan count 735, logical reads 2677, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 162, logical reads 380, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 81, logical reads 162, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 93, logical reads 186, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
```

Old query run 2:

```
Started executing query at Line 5
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
(21 rows affected)
Table 'Worktable'. Scan count 736, logical reads 2683, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 162, logical reads 380, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 81, logical reads 162, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 93, logical reads 186, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
```

New query run 1:

```
Started executing query at Line 55
SQL Server parse and compile time: 
   CPU time = 75 ms, elapsed time = 75 ms.
(44 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 160, logical reads 376, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 80, logical reads 160, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 85, logical reads 170, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
```

New query run 2:
```
Started executing query at Line 54
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
(24 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 160, logical reads 376, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 80, logical reads 160, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 85, logical reads 170, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 16 ms,  elapsed time = 15 ms.
Total execution time: 00:00:00.276
```

Would it be too much to ask that you perform the same test on one of your databases, preferably one with lots of statements in cache? This SQL is one of the heaviest in the agent, so just wanted to make sure that the change doesn't have adversary effects. 

The test consists of turning the diagnostics on in SQL Server Management Studio:

```
set statistics IO ON
set statistics time on
```

And then run the following two statements two times each.

Old query:

```
with qstats as (
    select query_hash, query_plan_hash, last_execution_time, last_elapsed_time,
            CONCAT(
                CONVERT(binary(64), plan_handle),
                CONVERT(binary(4), statement_start_offset),
                CONVERT(binary(4), statement_end_offset)) as plan_handle_and_offsets_order,
            CONCAT(
                CONVERT(VARCHAR(64), CONVERT(binary(64), plan_handle), 1),
                CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_start_offset), 1),
                CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_end_offset), 1)
                ) as plan_handle_and_offsets_varchar,
           (select value from sys.dm_exec_plan_attributes(plan_handle) where attribute = 'dbid') as dbid,
            execution_count, total_worker_time as total_worker_time, total_physical_reads, total_logical_writes, total_logical_reads, total_clr_time, total_elapsed_time, total_rows, total_dop, total_grant_kb, total_used_grant_kb, total_ideal_grant_kb, total_reserved_threads, total_used_threads, total_columnstore_segment_reads, total_columnstore_segment_skips, total_spills
    from sys.dm_exec_query_stats
),
qstats_aggr_order as (
    select FIRST_VALUE(plan_handle_and_offsets_varchar)
        over (partition by query_hash, query_plan_hash, S.dbid
            order by plan_handle_and_offsets_order desc) as plan_handle_and_offsets,
        S.*
    from qstats as S
),
qstats_aggr as (
    select query_hash, query_plan_hash, CAST(S.dbid as int) as dbid,
       D.name as database_name, max(plan_handle_and_offsets) as plan_handle_and_offsets,
       max(last_execution_time) as last_execution_time,
       max(last_elapsed_time) as last_elapsed_time,
       sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads, sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time, sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb, sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads, sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills
    from qstats_aggr_order S
    left join sys.databases D on S.dbid = D.database_id
    group by query_hash, query_plan_hash, S.dbid, D.name
),
qstats_aggr_split as (select TOP 10000
    convert(varbinary(64), convert(binary(64), substring(plan_handle_and_offsets, 1, 64), 1)) as plan_handle,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+1, 10), 1)) as statement_start_offset,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+11, 10), 1)) as statement_end_offset,
    * from qstats_aggr
    where DATEADD(ms, last_elapsed_time / 1000, last_execution_time) > dateadd(second, -60, getdate())
)
select
    SUBSTRING(text, (statement_start_offset / 2) + 1,
    ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset END
            - statement_start_offset) / 2) + 1) AS statement_text,
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt;
```

New query:

```
with qstats as (
    select query_hash, query_plan_hash, last_execution_time, last_elapsed_time,
            CONCAT(
                CONVERT(binary(64), plan_handle),
                CONVERT(binary(4), statement_start_offset),
                CONVERT(binary(4), statement_end_offset)) as plan_handle_and_offsets,
           (select value from sys.dm_exec_plan_attributes(plan_handle) where attribute = 'dbid') as dbid,
                       execution_count, total_worker_time as total_worker_time, total_physical_reads, total_logical_writes, total_logical_reads, total_clr_time, total_elapsed_time, total_rows, total_dop, total_grant_kb, total_used_grant_kb, total_ideal_grant_kb, total_reserved_threads, total_used_threads, total_columnstore_segment_reads, total_columnstore_segment_skips, total_spills
    from sys.dm_exec_query_stats
),
qstats_aggr as (
    select query_hash, query_plan_hash, CAST(S.dbid as int) as dbid,
       D.name as database_name, max(plan_handle_and_offsets) as plan_handle_and_offsets,
       max(last_execution_time) as last_execution_time,
       max(last_elapsed_time) as last_elapsed_time,
    sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads, sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time, sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb, sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads, sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills
    from qstats S
    left join sys.databases D on S.dbid = D.database_id
    group by query_hash, query_plan_hash, S.dbid, D.name
),
qstats_aggr_split as (select TOP 10000
    convert(varbinary(64), substring(plan_handle_and_offsets, 1, 64)) as plan_handle,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+1, 4))) as statement_start_offset,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+6, 4))) as statement_end_offset,
    * from qstats_aggr
    where DATEADD(ms, last_elapsed_time / 1000, last_execution_time) > dateadd(second, -60, getdate())
)
select
    SUBSTRING(text, (statement_start_offset / 2) + 1,
    ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset END
            - statement_start_offset) / 2) + 1) AS statement_text,
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt
```

If it's possible for you to do this test, please could you post the results here? They are in the tab "Messages".

Further, there's another similar query - stored in the variable `STATEMENT_METRICS_QUERY_NO_AGGREGATES`, which probably suffers the same issue. I don't know if you are affected by this. If not, we can do the changes.

it-ito commented:
@nenadnoveljic 
>> If the possibility of changing the results is acceptable, the test case can be modified.

> We can adapt the test cases to accommodate for the changes in the query. I think we can also add a grant to dbcc freeproccache to the datadog account when setting up the test database.

OK!
I fixed test cases(7c67cdc872d4b22b50080a14c558d4d31b09a95d , a14f8ece045f7105dcb61b1f18e8ee3f35ecf05b).
And I had committed poor performance fix for the test case so I reverted that (d550d430a5e1da6ac80a497bc1ca5c4632c99d6f).

> Would it be too much to ask that you perform the same test on one of your databases, preferably one with lots of statements in cache? This SQL is one of the heaviest in the agent, so just wanted to make sure that the change doesn't have adversary effects.

OK!!
My results are as follows.

# STATEMENT_METRICS_QUERY

## Old query (before this PR)

```sql
set statistics IO ON
set statistics time on
;

with qstats as (
    select query_hash, query_plan_hash, last_execution_time, last_elapsed_time,
            CONCAT(
                CONVERT(binary(64), plan_handle),
                CONVERT(binary(4), statement_start_offset),
                CONVERT(binary(4), statement_end_offset)) as plan_handle_and_offsets,
           (select value from sys.dm_exec_plan_attributes(plan_handle) where attribute = 'dbid') as dbid,
                       execution_count, total_worker_time as total_worker_time, total_physical_reads, total_logical_writes, total_logical_reads, total_clr_time, total_elapsed_time, total_rows, total_dop, total_grant_kb, total_used_grant_kb, total_ideal_grant_kb, total_reserved_threads, total_used_threads, total_columnstore_segment_reads, total_columnstore_segment_skips, total_spills
    from sys.dm_exec_query_stats
),
qstats_aggr as (
    select query_hash, query_plan_hash, CAST(S.dbid as int) as dbid,
       D.name as database_name, max(plan_handle_and_offsets) as plan_handle_and_offsets,
       max(last_execution_time) as last_execution_time,
       max(last_elapsed_time) as last_elapsed_time,
    sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads, sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time, sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb, sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads, sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills
    from qstats S
    left join sys.databases D on S.dbid = D.database_id
    group by query_hash, query_plan_hash, S.dbid, D.name
),
qstats_aggr_split as (select TOP 10000
    convert(varbinary(64), substring(plan_handle_and_offsets, 1, 64)) as plan_handle,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+1, 4))) as statement_start_offset,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+6, 4))) as statement_end_offset,
    * from qstats_aggr
    where DATEADD(ms, last_elapsed_time / 1000, last_execution_time) > dateadd(second, -60, getdate())
)
select
/* this statement is comment out because it cause an error before fix (https://github.com/DataDog/integrations-core/issues/14006)
    SUBSTRING(text, (statement_start_offset / 2) + 1,
    ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset END
            - statement_start_offset) / 2) + 1) AS statement_text,
*/
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt
```

Old query run 1:

```
SQL Server parse and compile time: 
   CPU time = 63 ms, elapsed time = 70 ms.
(851 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 14, logical reads 28, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 7, logical reads 14, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 1, logical reads 2, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 1516 ms,  elapsed time = 2174 ms.
```

Old query run 2:

```
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
(760 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 14, logical reads 28, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 7, logical reads 14, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 1, logical reads 2, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 1515 ms,  elapsed time = 2028 ms.
```

## New Query(after this PR)
```sql
set statistics IO ON
set statistics time on
;

with qstats as (
    select query_hash, query_plan_hash, last_execution_time, last_elapsed_time,
            CONCAT(
                CONVERT(VARCHAR(64), CONVERT(binary(64), plan_handle), 1),
                CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_start_offset), 1),
                CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_end_offset), 1)) as plan_handle_and_offsets,
           (select value from sys.dm_exec_plan_attributes(plan_handle) where attribute = 'dbid') as dbid,
                       execution_count, total_worker_time as total_worker_time, total_physical_reads, total_logical_writes, total_logical_reads, total_clr_time, total_elapsed_time, total_rows, total_dop, total_grant_kb, total_used_grant_kb, total_ideal_grant_kb, total_reserved_threads, total_used_threads, total_columnstore_segment_reads, total_columnstore_segment_skips, total_spills
    from sys.dm_exec_query_stats
),
qstats_aggr as (
    select query_hash, query_plan_hash, CAST(S.dbid as int) as dbid,
       D.name as database_name, max(plan_handle_and_offsets) as plan_handle_and_offsets,
       max(last_execution_time) as last_execution_time,
       max(last_elapsed_time) as last_elapsed_time,
    sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads, sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time, sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb, sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads, sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills
    from qstats S
    left join sys.databases D on S.dbid = D.database_id
    group by query_hash, query_plan_hash, S.dbid, D.name
),
qstats_aggr_split as (select TOP 10000
    convert(varbinary(64), convert(binary(64), substring(plan_handle_and_offsets, 1, 64), 1)) as plan_handle,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+1, 10), 1)) as statement_start_offset,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+11, 10), 1)) as statement_end_offset,
    * from qstats_aggr
    where DATEADD(ms, last_elapsed_time / 1000, last_execution_time) > dateadd(second, -60, getdate())
)
select
/*
    SUBSTRING(text, (statement_start_offset / 2) + 1,
    ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset END
            - statement_start_offset) / 2) + 1) AS statement_text,
*/

    SUBSTRING(text, (statement_start_offset / 2) + 1,
    ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset END
            - statement_start_offset) / 2) + 1) AS statement_text,
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt

```

New query run 1:

```
SQL Server parse and compile time: 
   CPU time = 63 ms, elapsed time = 69 ms.
(628 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 14, logical reads 28, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 7, logical reads 14, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 1, logical reads 2, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 1453 ms,  elapsed time = 1877 ms.
```

New query run 2:

```
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
(640 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysobjvalues'. Scan count 14, logical reads 28, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'syssingleobjrefs'. Scan count 7, logical reads 14, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'sysdbreg'. Scan count 1, logical reads 2, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 1485 ms,  elapsed time = 1871 ms.
```




# STATEMENT_METRICS_QUERY_NO_AGGREGATES

## Old query (before this PR)

```sql
-- STATEMENT_METRICS_QUERY_NO_AGGREGATES before
set statistics IO ON
set statistics time on
;

with qstats_aggr as (
    select TOP 10000 query_hash, query_plan_hash,
        max(CONCAT(
            CONVERT(binary(64), plan_handle),
            CONVERT(binary(4), statement_start_offset),
            CONVERT(binary(4), statement_end_offset))) as plan_handle_and_offsets,
            sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads,
            sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time,
            sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb,
            sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads,
            sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills

        from sys.dm_exec_query_stats S
        where last_execution_time > dateadd(second, -60, getdate())
        group by query_hash, query_plan_hash
),
qstats_aggr_split as (select
    convert(varbinary(64), substring(plan_handle_and_offsets, 1, 64)) as plan_handle,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+1, 4))) as statement_start_offset,
    convert(int, convert(varbinary(4), substring(plan_handle_and_offsets, 64+6, 4))) as statement_end_offset,
    * from qstats_aggr
)
select
/* this statement is comment out because it cause an error before fix (https://github.com/DataDog/integrations-core/issues/14006)
    SUBSTRING(text, (statement_start_offset / 2) + 1,
        ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset
    END - statement_start_offset) / 2) + 1) AS statement_text,*/
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt
```

Old query run 1:

```
SQL Server parse and compile time: 
   CPU time = 11 ms, elapsed time = 11 ms.
(628 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 718 ms,  elapsed time = 1117 ms.
```

Old query run 2:

```
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 10 ms.
(639 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 750 ms,  elapsed time = 1092 ms.
```

## New query (after this PR)

```sql
-- STATEMENT_METRICS_QUERY_NO_AGGREGATES after
set statistics IO ON
set statistics time on
;

with qstats_aggr as (
    select TOP 10000 query_hash, query_plan_hash,
        max(CONCAT(
            CONVERT(VARCHAR(64), CONVERT(binary(64), plan_handle), 1),
            CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_start_offset), 1),
            CONVERT(VARCHAR(10), CONVERT(varbinary(4), statement_end_offset), 1))) as plan_handle_and_offsets,
            sum(execution_count) as execution_count, sum(total_worker_time) as total_worker_time, sum(total_physical_reads) as total_physical_reads,
            sum(total_logical_writes) as total_logical_writes, sum(total_logical_reads) as total_logical_reads, sum(total_clr_time) as total_clr_time, sum(total_elapsed_time) as total_elapsed_time, 
            sum(total_rows) as total_rows, sum(total_dop) as total_dop, sum(total_grant_kb) as total_grant_kb, sum(total_used_grant_kb) as total_used_grant_kb, sum(total_ideal_grant_kb) as total_ideal_grant_kb, 
            sum(total_reserved_threads) as total_reserved_threads, sum(total_used_threads) as total_used_threads, sum(total_columnstore_segment_reads) as total_columnstore_segment_reads,
            sum(total_columnstore_segment_skips) as total_columnstore_segment_skips, sum(total_spills) as total_spills

        from sys.dm_exec_query_stats S
        where last_execution_time > dateadd(second, -60, getdate())
        group by query_hash, query_plan_hash
),
qstats_aggr_split as (select
    convert(varbinary(64), convert(binary(64), substring(plan_handle_and_offsets, 1, 64), 1)) as plan_handle,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+1, 10), 1)) as statement_start_offset,
    convert(int, convert(varbinary(10), substring(plan_handle_and_offsets, 64+11, 10), 1)) as statement_end_offset,
    * from qstats_aggr
)
select
/*
    SUBSTRING(text, (statement_start_offset / 2) + 1,
        ((CASE statement_end_offset
        WHEN -1 THEN DATALENGTH(text)
        ELSE statement_end_offset
    END - statement_start_offset) / 2) + 1) AS statement_text,*/
    qt.text,
    encrypted as is_encrypted,
    * from qstats_aggr_split
    cross apply sys.dm_exec_sql_text(plan_handle) qt
```


New query run 1:
```
SQL Server parse and compile time: 
   CPU time = 10 ms, elapsed time = 10 ms.
(782 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 719 ms,  elapsed time = 1185 ms.
```

New query run 2:

```
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
(797 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Workfile'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

 SQL Server Execution Times:
   CPU time = 781 ms,  elapsed time = 1204 ms.
```

it-ito commented on sqlserver/tests/test_statements.py:
```diff
@@ -452,9 +452,7 @@ def test_statement_metadata(
-    query = '''
-    -- Test comment
-    select * from sys.databases'''
+    query = 'select * from sys.databases'
```
### fix reason

When there is a cache of queries with the same `query_hash` and `query_plan_hash`, [this max(plan_handle_and_offsets)](https://github.com/DataDog/integrations-core/blob/adf5d2a523ca41945747eb858c092a0d39523b01/sqlserver/datadog_checks/sqlserver/statements.py#L68) may be different from before.

(1) and (2) are the same `query_hash` and `query_plan_hash`, and (2) was used preferentially before this fix.

(1)
```sql
-- Test comment
select * from sys.databases;
```

(2)
```sql
select * from sys.databases;
```

Because (2) is used preferentially in the test case, comments were not set on the following code.
https://github.com/DataDog/integrations-core/blob/a14f8ece045f7105dcb61b1f18e8ee3f35ecf05b/sqlserver/datadog_checks/sqlserver/statements.py#L310

As a result, the following code used metadata to set the comments

https://github.com/DataDog/integrations-core/blob/a14f8ece045f7105dcb61b1f18e8ee3f35ecf05b/sqlserver/datadog_checks/sqlserver/statements.py#L326-L327


On the other hand, (1) is used preferentially in the test case after fix, and `Test comment` set on the following code.
https://github.com/DataDog/integrations-core/blob/a14f8ece045f7105dcb61b1f18e8ee3f35ecf05b/sqlserver/datadog_checks/sqlserver/statements.py#L310

As a result, the following code didn't passed and failed assert with metadata.
https://github.com/DataDog/integrations-core/blob/a14f8ece045f7105dcb61b1f18e8ee3f35ecf05b/sqlserver/datadog_checks/sqlserver/statements.py#L326-L327

nenadnoveljic commented on sqlserver/tests/test_statements.py:
```diff
@@ -452,9 +452,7 @@ def test_statement_metadata(
 ):
     check = SQLServer(CHECK_NAME, {}, [dbm_instance])
 
-    query = '''
-    -- Test comment
-    select * from sys.databases'''
```
Can we perhaps keep the original query with the comment after adding the preceding `dbcc freeproccache` call?

it-ito replied:
No, in that case the test failed.

Before this PR, this query is not important.
For example, In the case of different comments (https://github.com/it-ito/integrations-core/commit/63d4ab8e291dc68bd90d898bdf139983df0b96fa)  or different table (https://github.com/it-ito/integrations-core/commit/228a9dd3897c99822d84c67fe125af66c6cb9e2d) , test case succeed before. 

Perhaps there was a `select * from sys.databases` running somewhere that was being used.

nenadnoveljic replied:
I see, thanks. The query is being used by Agent for auto-discovery and it's possible that it interferes with the test case.

But do we then need `dbcc freeproccache` call at all?

Does it work if we remove `dbcc freeproccache` and replace `sys.databases` with `sys.sysusers`, as the latter isn't queried by the agent? That would be the perfect solution because it would both fix the original problem and preserve the test case with the comment. If that would be too much to ask, we can go further with your changes and later deal with the test case problem internally. 

Thanks for your help!
